Kernel: Building the kernel with LTO doesn't work

When I tried to build the kernel with LTO enabled, it gave me these errors:
```
/home/cleverking/serenity-os/Toolchain/Local/i686/lib/gcc/i686-pc-serenity/11.1.0/../../../../i686-pc-serenity/bin/ld: /tmp/ccts649E.ltrans0.ltrans.o: in function `illegal_instruction_asm_entry':
/home/cleverking/serenity-os/Build/i686/../.././Kernel/Interrupts/IRQHandler.h:86: undefined reference to `enter_trap_no_irq'
/home/cleverking/serenity-os/Toolchain/Local/i686/lib/gcc/i686-pc-serenity/11.1.0/../../../../i686-pc-serenity/bin/ld: /tmp/ccts649E.ltrans0.ltrans.o: in function `divide_error_asm_entry':
/home/cleverking/serenity-os/Build/i686/../.././Kernel/Interrupts/IRQHandler.h:108: undefined reference to `enter_trap_no_irq'
/home/cleverking/serenity-os/Toolchain/Local/i686/lib/gcc/i686-pc-serenity/11.1.0/../../../../i686-pc-serenity/bin/ld: /tmp/ccts649E.ltrans0.ltrans.o: in function `general_protection_fault_asm_entry':
/home/cleverking/serenity-os/Build/i686/../.././Kernel/Interrupts/IRQHandler.h:129: undefined reference to `enter_trap_no_irq'
/home/cleverking/serenity-os/Toolchain/Local/i686/lib/gcc/i686-pc-serenity/11.1.0/../../../../i686-pc-serenity/bin/ld: /tmp/ccts649E.ltrans0.ltrans.o: in function `fpu_exception_asm_entry':
/home/cleverking/serenity-os/Build/i686/../.././Kernel/Interrupts/IRQHandler.h:151: undefined reference to `enter_trap_no_irq'
/home/cleverking/serenity-os/Toolchain/Local/i686/lib/gcc/i686-pc-serenity/11.1.0/../../../../i686-pc-serenity/bin/ld: /tmp/ccts649E.ltrans0.ltrans.o: in function `page_fault_asm_entry':
/home/cleverking/serenity-os/Build/i686/../.././Kernel/Interrupts/IRQHandler.h:172: undefined reference to `enter_trap_no_irq'
/home/cleverking/serenity-os/Toolchain/Local/i686/lib/gcc/i686-pc-serenity/11.1.0/../../../../i686-pc-serenity/bin/ld: /tmp/ccts649E.ltrans0.ltrans.o:/home/cleverking/serenity-os/Build/i686/../.././Kernel/Interrupts/IRQHandler.h:194: more undefined references to `enter_trap_no_irq' follow
```
Disabling LTO fixed the build.

Sorry, forgot a part of the build log
```
../../../../x86_64-pc-serenity/bin/ld: DWARF error: invalid abstract instance DIE ref
/tmp/ccS1IlFB.ltrans0.ltrans.o: in function `illegal_instruction_asm_entry':
<artificial>:(.text+0xc3): undefined reference to `illegal_instruction_handler'
<artificial>:(.text+0xc3): relocation truncated to fit: R_X86_64_PLT32 against undefined symbol `illegal_instruction_handler'
/home/cleverking/serenity-os/Toolchain/Local/x86_64/lib/gcc/x86_64-pc-serenity/11.1.0/../../../../x86_64-pc-serenity/bin/ld: /tmp/ccS1IlFB.ltrans0.ltrans.o: in function `divide_error_asm_entry':
<artificial>:(.text+0xf8): undefined reference to `divide_error_handler'
<artificial>:(.text+0xf8): relocation truncated to fit: R_X86_64_PLT32 against undefined symbol `divide_error_handler'
/home/cleverking/serenity-os/Toolchain/Local/x86_64/lib/gcc/x86_64-pc-serenity/11.1.0/../../../../x86_64-pc-serenity/bin/ld: /tmp/ccS1IlFB.ltrans0.ltrans.o: in function `general_protection_fault_asm_entry':
<artificial>:(.text+0x133): undefined reference to `general_protection_fault_handler'
<artificial>:(.text+0x133): relocation truncated to fit: R_X86_64_PLT32 against undefined symbol `general_protection_fault_handler'
/home/cleverking/serenity-os/Toolchain/Local/x86_64/lib/gcc/x86_64-pc-serenity/11.1.0/../../../../x86_64-pc-serenity/bin/ld: /tmp/ccS1IlFB.ltrans0.ltrans.o: in function `fpu_exception_asm_entry':
<artificial>:(.text+0x16c): undefined reference to `fpu_exception_handler'
<artificial>:(.text+0x16c): relocation truncated to fit: R_X86_64_PLT32 against undefined symbol `fpu_exception_handler'/home/cleverking/serenity-os/Toolchain/Local/x86_64/lib/gcc/x86_64-pc-serenity/11.1.0/../../../../x86_64-pc-serenity/bin/ld: /tmp/ccS1IlFB.ltrans0.ltrans.o: in function `page_fault_asm_entry':
<artificial>:(.text+0x1a7): undefined reference to `page_fault_handler'
<artificial>:(.text+0x1a7): relocation truncated to fit: R_X86_64_PLT32 against undefined symbol `page_fault_handler'
/home/cleverking/serenity-os/Toolchain/Local/x86_64/lib/gcc/x86_64-pc-serenity/11.1.0/../../../../x86_64-pc-serenity/bin/ld: /tmp/ccS1IlFB.ltrans0.ltrans.o: in function `debug_asm_entry':
<artificial>:(.text+0x1e0): undefined reference to `debug_handler'
<artificial>:(.text+0x1e0): relocation truncated to fit: R_X86_64_PLT32 against undefined symbol `debug_handler'
/home/cleverking/serenity-os/Toolchain/Local/x86_64/lib/gcc/x86_64-pc-serenity/11.1.0/../../../../x86_64-pc-serenity/bin/ld: /tmp/ccS1IlFB.ltrans0.ltrans.o: in function `breakpoint_asm_entry':
<artificial>:(.text+0x215): undefined reference to `breakpoint_handler'
<artificial>:(.text+0x215): relocation truncated to fit: R_X86_64_PLT32 against undefined symbol `breakpoint_handler'
/home/cleverking/serenity-os/Toolchain/Local/x86_64/lib/gcc/x86_64-pc-serenity/11.1.0/../../../../x86_64-pc-serenity/bin/ld: /home/cleverking/serenity-os/Toolchain/Local/x86_64/lib/gcc/x86_64-pc-serenity/11.1.0/../../../../x86_64-pc-serenity/bin/ld: DWARF error: invalid abstract instance DIE ref
/tmp/ccS1IlFB.ltrans19.ltrans.o: in function `Kernel::Process::initialize()':
<artificial>:(.unmap_after_init+0x611): undefined reference to `asm_signal_trampoline'
/home/cleverking/serenity-os/Toolchain/Local/x86_64/lib/gcc/x86_64-pc-serenity/11.1.0/../../../../x86_64-pc-serenity/bin/ld: <artificial>:(.unmap_after_init+0x639): undefined reference to `asm_signal_trampoline_end'
collect2: error: ld returned 1 exit status
```

--- a/Kernel/Arch/x86/TrapFrame.h
+++ b/Kernel/Arch/x86/TrapFrame.h
@@ -34,7 +34,7 @@ struct TrapFrame {
 
 static_assert(TRAP_FRAME_SIZE == sizeof(TrapFrame));
 
-extern "C" void enter_trap_no_irq(TrapFrame* trap);
+extern "C" void enter_trap_no_irq(TrapFrame* trap) __attribute__((used));
 extern "C" void enter_trap(TrapFrame*) __attribute__((used));
 extern "C" void exit_trap(TrapFrame*) __attribute__((used));
 

--- a/Kernel/Arch/x86/common/Interrupts.cpp
+++ b/Kernel/Arch/x86/common/Interrupts.cpp
@@ -94,69 +94,69 @@ static EntropySource s_entropy_source_interrupts { EntropySource::Static::Interr
         "    jmp common_trap_exit \n");
 
 #elif ARCH(X86_64)
-#define EH_ENTRY(ec, title)                         \
-    extern "C" void title##_asm_entry();            \
-    extern "C" void title##_handler(TrapFrame*);    \
-    asm(                                            \
-        ".globl " #title "_asm_entry\n"             \
-        "" #title "_asm_entry: \n"                  \
-        "    pushq %r15\n"                          \
-        "    pushq %r14\n"                          \
-        "    pushq %r13\n"                          \
-        "    pushq %r12\n"                          \
-        "    pushq %r11\n"                          \
-        "    pushq %r10\n"                          \
-        "    pushq %r9\n"                           \
-        "    pushq %r8\n"                           \
-        "    pushq %rax\n"                          \
-        "    pushq %rcx\n"                          \
-        "    pushq %rdx\n"                          \
-        "    pushq %rbx\n"                          \
-        "    pushq %rsp\n"                          \
-        "    pushq %rbp\n"                          \
-        "    pushq %rsi\n"                          \
-        "    pushq %rdi\n"                          \
-        "    pushq %rsp \n" /* set TrapFrame::regs */ \
-        "    subq $" __STRINGIFY(TRAP_FRAME_SIZE - 8) ", %rsp \n" \
-        "    subq $0x8, %rsp\n" /* align stack */   \
-        "    lea 0x8(%rsp), %rdi \n"                \
-        "    cld\n"                                 \
-        "    call enter_trap_no_irq \n"             \
-        "    lea 0x8(%rsp), %rdi \n"                \
-        "    call " #title "_handler\n"             \
-        "    addq $0x8, %rsp\n" /* undo alignment */\
+#define EH_ENTRY(ec, title)                                               \
+    extern "C" void title##_asm_entry();                                  \
+    extern "C" void title##_handler(TrapFrame*) __attribute__((used));    \
+    asm(                                                                  \
+        ".globl " #title "_asm_entry\n"                                   \
+        "" #title "_asm_entry: \n"                                        \
+        "    pushq %r15\n"                                                \
+        "    pushq %r14\n"                                                \
+        "    pushq %r13\n"                                                \
+        "    pushq %r12\n"                                                \
+        "    pushq %r11\n"                                                \
+        "    pushq %r10\n"                                                \
+        "    pushq %r9\n"                                                 \
+        "    pushq %r8\n"                                                 \
+        "    pushq %rax\n"                                                \
+        "    pushq %rcx\n"                                                \
+        "    pushq %rdx\n"                                                \
+        "    pushq %rbx\n"                                                \
+        "    pushq %rsp\n"                                                \
+        "    pushq %rbp\n"                                                \
+        "    pushq %rsi\n"                                                \
+        "    pushq %rdi\n"                                                \
+        "    pushq %rsp \n" /* set TrapFrame::regs */                     \
+        "    subq $" __STRINGIFY(TRAP_FRAME_SIZE - 8) ", %rsp \n"         \
+        "    subq $0x8, %rsp\n" /* align stack */                         \
+        "    lea 0x8(%rsp), %rdi \n"                                      \
+        "    cld\n"                                                       \
+        "    call enter_trap_no_irq \n"                                   \
+        "    lea 0x8(%rsp), %rdi \n"                                      \
+        "    call " #title "_handler\n"                                   \
+        "    addq $0x8, %rsp\n" /* undo alignment */                      \
         "    jmp common_trap_exit \n");
 
-#define EH_ENTRY_NO_CODE(ec, title)                 \
-    extern "C" void title##_handler(TrapFrame*);    \
-    extern "C" void title##_asm_entry();            \
-asm(                                                \
-        ".globl " #title "_asm_entry\n"             \
-        "" #title "_asm_entry: \n"                  \
-        "    pushq $0x0\n"                          \
-        "    pushq %r15\n"                          \
-        "    pushq %r14\n"                          \
-        "    pushq %r13\n"                          \
-        "    pushq %r12\n"                          \
-        "    pushq %r11\n"                          \
-        "    pushq %r10\n"                          \
-        "    pushq %r9\n"                           \
-        "    pushq %r8\n"                           \
-        "    pushq %rax\n"                          \
-        "    pushq %rcx\n"                          \
-        "    pushq %rdx\n"                          \
-        "    pushq %rbx\n"                          \
-        "    pushq %rsp\n"                          \
-        "    pushq %rbp\n"                          \
-        "    pushq %rsi\n"                          \
-        "    pushq %rdi\n"                          \
-        "    pushq %rsp \n" /* set TrapFrame::regs */ \
-        "    subq $" __STRINGIFY(TRAP_FRAME_SIZE - 8) ", %rsp \n" \
-        "    movq %rsp, %rdi \n"                    \
-        "    cld\n"                                 \
-        "    call enter_trap_no_irq \n"             \
-        "    movq %rsp, %rdi \n"                    \
-        "    call " #title "_handler\n"             \
+#define EH_ENTRY_NO_CODE(ec, title)                                       \
+    extern "C" void title##_handler(TrapFrame*) __attribute__((used));    \
+    extern "C" void title##_asm_entry();                                  \
+asm(                                                                      \
+        ".globl " #title "_asm_entry\n"                                   \
+        "" #title "_asm_entry: \n"                                        \
+        "    pushq $0x0\n"                                                \
+        "    pushq %r15\n"                                                \
+        "    pushq %r14\n"                                                \
+        "    pushq %r13\n"                                                \
+        "    pushq %r12\n"                                                \
+        "    pushq %r11\n"                                                \
+        "    pushq %r10\n"                                                \
+        "    pushq %r9\n"                                                 \
+        "    pushq %r8\n"                                                 \
+        "    pushq %rax\n"                                                \
+        "    pushq %rcx\n"                                                \
+        "    pushq %rdx\n"                                                \
+        "    pushq %rbx\n"                                                \
+        "    pushq %rsp\n"                                                \
+        "    pushq %rbp\n"                                                \
+        "    pushq %rsi\n"                                                \
+        "    pushq %rdi\n"                                                \
+        "    pushq %rsp \n" /* set TrapFrame::regs */                     \
+        "    subq $" __STRINGIFY(TRAP_FRAME_SIZE - 8) ", %rsp \n"         \
+        "    movq %rsp, %rdi \n"                                          \
+        "    cld\n"                                                       \
+        "    call enter_trap_no_irq \n"                                   \
+        "    movq %rsp, %rdi \n"                                          \
+        "    call " #title "_handler\n"                                   \
         "    jmp common_trap_exit \n");
 #endif
 

